File: js/phase-analysis.js
// Phase Analysis System
// (Same logic as before, preserved completely)

function analyzeTracePhases(steps) {
  if (!steps || steps.length === 0) return [];
  
  const phases = [];
  let currentPhase = { 
    type: 'unknown', 
    startIdx: 0, 
    addresses: new Set(),
    startAmount: parseFloat(sompiToKAS(steps[0]?.amount || 0)), 
    exchangeName: null 
  };
  
  for (let i = 0; i < steps.length; i++) {
    const step = steps[i];
    const amount = parseFloat(sompiToKAS(step.amount || 0));
    const addr = step.address;
    const isCoinbase = step.type === 'coinbase' || step.isCoinbase;
    const exchangeName = getExchangeNameFromAddress(addr);
    const isExchange = isExchangeLike(addr, amount);
    
    let stepPhaseType;
    
    if (isCoinbase) { 
      stepPhaseType = 'mining'; 
    } else if (isExchange) { 
      stepPhaseType = 'exchange'; 
    } else if (amount < 100 && currentPhase.type === 'exchange') { 
      stepPhaseType = 'distribution'; 
    } else if (amount < 1000 && !isExchange) { 
      stepPhaseType = 'consolidation'; 
    } else { 
      stepPhaseType = 'exchange'; 
    }
    
    if (i === 0) {
      currentPhase.type = stepPhaseType;
      currentPhase.exchangeName = exchangeName;
    } else {
      const typeChanged = stepPhaseType !== currentPhase.type;
      const exchangeChanged = isExchange && exchangeName && exchangeName !== currentPhase.exchangeName;
      const enteringExchange = stepPhaseType === 'exchange' && currentPhase.type !== 'exchange';
      const leavingExchange = stepPhaseType === 'distribution' && currentPhase.type === 'exchange';
      const enteringMining = stepPhaseType === 'mining' && currentPhase.type !== 'mining';
      
      if (typeChanged || exchangeChanged || enteringExchange || leavingExchange || enteringMining) {
        phases.push({
          type: currentPhase.type,
          exchangeName: currentPhase.exchangeName,
          startIdx: currentPhase.startIdx,
          endIdx: i - 1,
          startAmount: currentPhase.startAmount,
          endAmount: parseFloat(sompiToKAS(steps[i - 1]?.amount || 0)),
          stepCount: i - currentPhase.startIdx,
          addresses: Array.from(currentPhase.addresses)
        });
        
        currentPhase = {
          type: stepPhaseType,
          startIdx: i,
          addresses: new Set([addr]),
          startAmount: amount,
          exchangeName: exchangeName
        };
      } else {
        currentPhase.addresses.add(addr);
        if (exchangeName && !currentPhase.exchangeName) {
          currentPhase.exchangeName = exchangeName;
        }
      }
    }
  }
  
  phases.push({
    type: currentPhase.type,
    exchangeName: currentPhase.exchangeName,
    startIdx: currentPhase.startIdx,
    endIdx: steps.length - 1,
    startAmount: currentPhase.startAmount,
    endAmount: parseFloat(sompiToKAS(steps[steps.length - 1]?.amount || 0)),
    stepCount: steps.length - currentPhase.startIdx,
    addresses: Array.from(currentPhase.addresses)
  });
  
  return phases;
}

function getExchangeNameFromAddress(addr) {
  return lookupKnownAddress(addr)?.name || null;
}

function getPhaseLabel(type, exchangeName) {
  const labels = {
    'mining': '⛏️ Mining',
    'consolidation': '🔄 Consolidation',
    'exchange': exchangeName ? `🏦 ${exchangeName}` : '🏦 Exchange Activity',
    'distribution': '📤 Distribution',
    'unknown': '❓ Unknown'
  };
  return labels[type] || type;
}

function getPhaseClass(type) {
  return `phase-${type}`;
}

// Get simplified path for coin story
function getStoryPath(phases) {
  return phases.map(phase => ({
    type: phase.type,
    label: getPhaseLabel(phase.type, phase.exchangeName),
    exchangeName: phase.exchangeName,
    icon: getPhaseIcon(phase.type),
    cssClass: getPhaseNodeClass(phase.type)
  }));
}

function getPhaseIcon(type) {
  const icons = {
    'mining': '⛏️',
    'consolidation': '🔄',
    'exchange': '🏦',
    'distribution': '📤',
    'unknown': '❓'
  };
  return icons[type] || '❓';
}

function getPhaseNodeClass(type) {
  const classes = {
    'mining': 'mining',
    'exchange': 'exchange',
    'consolidation': 'personal',
    'distribution': 'personal'
  };
  return classes[type] || '';
}

// Calculate trace confidence
function calculateConfidence(traceResult) {
  if (!traceResult || traceResult.error) return 0;
  
  let confidence = 0;
  const weights = UI_CONFIG.confidenceWeights;
  
  if (traceResult.foundCoinbase) confidence += weights.coinbaseFound;
  
  const hasKnownExchange = traceResult.steps?.some(s => 
    s.isExchangeAddress || s.isBehavioralExchange
  );
  if (hasKnownExchange) confidence += weights.exchangeIdentified;
  
  if (traceResult.steps?.length > 0 && !traceResult.steps.some(s => s.type === 'dead_end')) {
    confidence += weights.completePath;
  }
  
  if (!traceResult.validationWarnings?.length && !traceResult.invalidAddresses?.length) {
    confidence += weights.noDeadEnds;
  }
  
  return Math.min(100, confidence);
}

function getConfidenceLabel(confidence) {
  if (confidence >= 90) return 'High';
  if (confidence >= 60) return 'Medium';
  if (confidence >= 30) return 'Low';
  return 'Minimal';
}

File: backend/phase_analysis.py
"""Phase analysis for coin tracing."""

from backend.utils import sompi_to_kas
from backend.behavioral import is_exchange_like
from backend.known_addresses import lookup_known_address


def analyze_trace_phases(steps: list) -> list:
    """Analyze trace steps and categorize into phases."""
    if not steps:
        return []
    
    phases = []
    current_phase = {
        "type": "unknown",
        "start_idx": 0,
        "addresses": set(),
        "start_amount": float(sompi_to_kas(steps[0].get("amount", 0))),
        "exchange_name": None
    }
    
    for i, step in enumerate(steps):
        amount = float(sompi_to_kas(step.get("amount", 0)))
        addr = step.get("address", "")
        is_coinbase = step.get("type") == "coinbase" or step.get("is_coinbase", False)
        exchange_name = _get_exchange_name(addr)
        is_exchange = is_exchange_like(addr, amount)
        
        # Determine step phase type
        if is_coinbase:
            step_phase = "mining"
        elif is_exchange:
            step_phase = "exchange"
        elif amount < 100 and current_phase["type"] == "exchange":
            step_phase = "distribution"
        elif amount < 1000 and not is_exchange:
            step_phase = "consolidation"
        else:
            step_phase = "exchange"
        
        if i == 0:
            current_phase["type"] = step_phase
            current_phase["exchange_name"] = exchange_name
        else:
            type_changed = step_phase != current_phase["type"]
            exchange_changed = (
                is_exchange and 
                exchange_name and 
                exchange_name != current_phase["exchange_name"]
            )
            entering_exchange = step_phase == "exchange" and current_phase["type"] != "exchange"
            leaving_exchange = step_phase == "distribution" and current_phase["type"] == "exchange"
            entering_mining = step_phase == "mining" and current_phase["type"] != "mining"
            
            if type_changed or exchange_changed or entering_exchange or leaving_exchange or entering_mining:
                phases.append({
                    "type": current_phase["type"],
                    "exchange_name": current_phase["exchange_name"],
                    "start_idx": current_phase["start_idx"],
                    "end_idx": i - 1,
                    "start_amount": current_phase["start_amount"],
                    "end_amount": float(sompi_to_kas(steps[i-1].get("amount", 0))),
                    "step_count": i - current_phase["start_idx"],
                    "addresses": list(current_phase["addresses"])
                })
                
                current_phase = {
                    "type": step_phase,
                    "start_idx": i,
                    "addresses": {addr},
                    "start_amount": amount,
                    "exchange_name": exchange_name
                }
            else:
                current_phase["addresses"].add(addr)
                if exchange_name and not current_phase["exchange_name"]:
                    current_phase["exchange_name"] = exchange_name
    
    # Add final phase
    phases.append({
        "type": current_phase["type"],
        "exchange_name": current_phase["exchange_name"],
        "start_idx": current_phase["start_idx"],
        "end_idx": len(steps) - 1,
        "start_amount": current_phase["start_amount"],
        "end_amount": float(sompi_to_kas(steps[-1].get("amount", 0))),
        "step_count": len(steps) - current_phase["start_idx"],
        "addresses": list(current_phase["addresses"])
    })
    
    return phases


def _get_exchange_name(addr: str) -> str | None:
    """Get exchange name from known addresses."""
    info = lookup_known_address(addr)
    return info["name"] if info else None


def get_phase_label(phase_type: str, exchange_name: str | None = None) -> str:
    """Get human-readable label for a phase type."""
    labels = {
        "mining": "⛏️ Mining",
        "consolidation": "🔄 Consolidation",
        "exchange": f"🏦 {exchange_name}" if exchange_name else "🏦 Exchange Activity",
        "distribution": "📤 Distribution",
        "unknown": "❓ Unknown"
    }
    return labels.get(phase_type, phase_type)


def get_phase_icon(phase_type: str) -> str:
    """Get icon for a phase type."""
    icons = {
        "mining": "⛏️",
        "consolidation": "🔄",
        "exchange": "🏦",
        "distribution": "📤",
        "unknown": "❓"
    }
    return icons.get(phase_type, "❓")


def get_story_path(phases: list) -> list:
    """Get simplified path for coin story display."""
    return [
        {
            "type": p["type"],
            "label": get_phase_label(p["type"], p.get("exchange_name")),
            "exchange_name": p.get("exchange_name"),
            "icon": get_phase_icon(p["type"])
        }
        for p in phases
    ]


def calculate_confidence(trace_result: dict) -> int:
    """Calculate confidence score for a trace result."""
    from backend.config import UI_CONFIG
    
    if not trace_result or trace_result.get("error"):
        return 0
    
    weights = UI_CONFIG["confidence_weights"]
    confidence = 0
    steps = trace_result.get("steps", [])
    
    if trace_result.get("found_coinbase"):
        confidence += weights["coinbase_found"]
    
    has_exchange = any(
        s.get("is_exchange_address") or s.get("is_behavioral_exchange")
        for s in steps
    )
    if has_exchange:
        confidence += weights["exchange_identified"]
    
    if steps and not any(s.get("type") == "dead_end" for s in steps):
        confidence += weights["complete_path"]
    
    if not trace_result.get("validation_warnings") and not trace_result.get("invalid_addresses"):
        confidence += weights["no_dead_ends"]
    
    return min(100, confidence)


def get_confidence_label(confidence: int) -> str:
    """Get label for confidence score."""
    if confidence >= 90:
        return "High"
    if confidence >= 60:
        return "Medium"
    if confidence >= 30:
        return "Low"
    return "Minimal"
